File: stack_queues/542_01_matrix.java
class Solution {
    
    private int[][] directions = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    
    public int[][] updateMatrix(int[][] mat) {
        int m = mat.length, n = mat[0].length;
        int[][] res = new int[m][n];
        boolean[][] seen = new boolean[m][n];
        Queue<int[]> queue = new LinkedList<>();
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    queue.offer(new int[] {i, j, 0}); 
                    seen[i][j] = true;
                }
            }          
        }
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int row = curr[0];
            int col = curr[1];
            int dist = curr[2];
            
            for (int[] direction : directions) {
                int nr = direction[0] + row;
                int nc = direction[1] + col;
                if (nr >= 0 && nc >= 0 && nr < m && nc < n && !seen[nr][nc]) {
                    seen[nr][nc] = true;
                    res[nr][nc] = dist + 1;
                    queue.offer(new int[] {nr, nc, dist + 1});
                }
            }
        }
        return res;
    }
}

File: zmisc/542_01_matrix.py
class Solution:
    def updateMatrix(self, mat: list[list[int]]) -> list[list[int]]:
        r, c = len(mat), len(mat[0])
        q = []
        for i in range(r):
            for j in range(c):
                if mat[i][j] == 0:
                    q.append((i, j))
                else:
                    mat[i][j] = '#'

        for row, column in q:
            for dx, dy in (1,0), (-1, 0), (0, 1), (0, -1):
                nr = row + dx
                nc = column + dy
                if 0 <= nr < r and 0 <= nc < c and mat[nr][nc] == '#':
                    mat[nr][nc] = mat[row][column] + 1
                    q.append((nr, nc))
        return mat


mat = [[0,0,0],[0,1,0],[0,0,0]]
mat = [[0,0,0],[0,1,0],[1,1,1]]
print(Solution().updateMatrix(mat))